src/ucp/core/ucp_listener.c

```diff
@@ -172,99 +172,239 @@ static void ucp_listener_conn_request_callback(uct_iface_h tl_iface, void *arg,
     ucp_worker_signal_internal(listener->worker);
 }
 
-ucs_status_t ucp_listener_query(ucp_listener_h listener, ucp_listener_attr_t *attr)
+static ucs_status_t
+ucp_listener_query_uct_port(ucp_listener_h listener,
+                            ucp_rsc_index_t uct_listener_idx, uint16_t *port_p)
+{
+    uct_listener_attr_t attr;
+    ucs_status_t status;
+
+    attr.field_mask = UCT_LISTENER_ATTR_FIELD_SOCKADDR;
+    status = uct_listener_query(listener->cms.listeners[uct_listener_idx],
+                                &attr);
+    if (status != UCS_OK) {
+        ucs_error("failed to query UCT listener %p status %s",
+                  listener->cms.listeners[uct_listener_idx],
+                  ucs_status_string(status));
+        return status;
+    }
+
+    return ucs_sockaddr_get_port((struct sockaddr *)&attr.sockaddr, port_p);
+}
+
+static ucs_status_t
+ucp_listener_uct_listeners_query_port(ucp_listener_h listener, int *port_p)
+{
+    uint16_t port_0 = 0, port_i;
+    int i;
+    ucs_status_t status;
+
+    if (listener->cms.num_listeners == 0) {
+        ucs_error("there are no UCT listeners");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    /* Make sure that all the listeners are listening on the same port */
+    for (i = 0; i < listener->cms.num_listeners; i++) {
+        status = ucp_listener_query_uct_port(listener, i, &port_i);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        if (i == 0) {
+            port_0 = port_i;
+        } else if (port_0 != port_i) {
+            ucs_error("different ports detected on the listener: %d and %d",
+                      port_0, port_i);
+            return UCS_ERR_IO_ERROR;
+        }
+    }
+
+    *port_p = port_0;
+    return UCS_OK;
+}
+
+static ucs_status_t ucp_listener_ifaces_query_port(ucp_listener_h listener,
+                                                   int *port_p)
 {
     int i, port;
 
-    ucs_assert(listener->num_wifaces > 0);
-    port = listener->wifaces[0].attr.listen_port;
+    ucs_assert(listener->ifaces.num_wifaces > 0);
+    port = listener->ifaces.wifaces[0].attr.listen_port;
 
     /* Make sure that all the listening sockaddr ifaces are listening on the same port */
-    for (i = 1; i < listener->num_wifaces; i++) {
-        if (port != listener->wifaces[i].attr.listen_port) {
+    for (i = 1; i < listener->ifaces.num_wifaces; i++) {
+        if (port != listener->ifaces.wifaces[i].attr.listen_port) {
```

Comment:
maybe ucp_listener_query_uct_port() should handle the difference between ifaces and cms. and the top-level function will go over all N ifaces/cms and just make sure all return the same port number

Reply:
See https://github.com/openucx/ucx/pull/4026#discussion_r313449325 , we have API limitation to unify all these things in a good way.